### DISCUSSION TITLE
Subscription products?

### DISCUSSION THREAD

**Discussion by binaryfire** [lunarphp/lunar#344] [Q&A]
Hi guys!

Do you plan to add support for subscription / recurring payment products?

Cheers

**Comment by glennjacobs** [comment]
It is something that we need to be considering, but it's not on our radar right now as we're focusing on the core functionality.

But I would be interested in hearing people's requirements for this.

**Comment by brockcaldwell** [comment]
I've been toying with Lunar to see if I can use it for client sites and it meets most of my requirements. The only thing missing is Subscription products. Has this been added to any Roadmap or is it still being considered? I would offer a PR but most of my free time is filled being a dad.

**Comment by JiveDig** [comment]
2026 and I'm in the same boat. Lunar has looked awesome, but not being able to offer subscriptions along side one-time payments is a blocker for us.

**Reply by glennjacobs** [reply]
It would be good to get a spec of what the Lunar core would need for subscription support.

**Reply by tazzytazzy** [reply]
I think Lunar can make subscriptions much easier without needing to become a full subscription system. Every company will have different lifecycle rules, billing policies, entitlements, and workflows, so Lunar could provide a few strong commerce seams while applications continue to own those business rules.

- **ProductVariant and Price:** represent sellable plans and add-ons with normal Lunar catalog and integer-price identities.
- **Cart-line pricing pipeline:** let an application select an exact price version instead of relying only on default price matching.
- **Cart validators:** enforce application-owned eligibility, compatibility, cycle, and quantity rules.
- **CartLine/OrderLine metadata:** preserve plan, offer, subscription, and add-on occurrence identities through Order creation.
- **Provider-free Order placement Action:** place an already-paid draft Order without invoking a payment provider, making initial and renewal Order finalization safer.
- **PriceValue and StorefrontContext:** support exact money calculations and non-session renewal or background-job workflows.

I can expand on these topics more if you're interested. We're finally just about to implement subscriptions and i'd be happy to work on these and generate some PRs against 2.x.
